// @unstoppabledomains/resolution/build/utils/Utils.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var index_1 = require("./index");
describe('Lodash', function () {
    describe('set', function () {
        it('should set new values', function () {
            expect(index_1.set({}, 'a.b', 1)).toStrictEqual({ a: { b: 1 } });
        });
        it('should set new property to the existing object', function () {
            expect(index_1.set({ a: 1 }, 'a.b', 1)).toStrictEqual({ a: { b: 1 } });
        });
        it('should work on deeper levels', function () {
            expect(index_1.set({ a: { b: 2 } }, 'a.b', 1)).toStrictEqual({ a: { b: 1 } });
        });
        it('should just work', function () {
            expect(index_1.set({}, 'crypto.BCH.address', 'qrq4sk49ayvepqz7j7ep8x4km2qp8lauvcnzhveyu6')).toStrictEqual({
                crypto: {
                    BCH: { address: 'qrq4sk49ayvepqz7j7ep8x4km2qp8lauvcnzhveyu6' },
                },
            });
        });
    });
    it('should invert the object', function () {
        var object = {
            mainnet: 'https://api.zilliqa.com',
            testnet: 'https://dev-api.zilliqa.com',
            localnet: 'http://localhost:4201',
        };
        expect(index_1.invert(object)).toStrictEqual({
            'https://api.zilliqa.com': 'mainnet',
            'https://dev-api.zilliqa.com': 'testnet',
            'http://localhost:4201': 'localnet',
        });
    });
});
