// @unstoppabledomains/resolution/build/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var NamingServiceName;
(function (NamingServiceName) {
    NamingServiceName["ENS"] = "ENS";
    NamingServiceName["CNS"] = "CNS";
    NamingServiceName["ZNS"] = "ZNS";
})(NamingServiceName = exports.NamingServiceName || (exports.NamingServiceName = {}));
exports.UDApiDefaultUrl = 'https://unstoppabledomains.com/api/v1';
exports.DefaultAPI = {
    url: exports.UDApiDefaultUrl,
};
var NullAddress;
(function (NullAddress) {
    NullAddress[NullAddress["0x"] = 0] = "0x";
    NullAddress[NullAddress["0x0000000000000000000000000000000000000000"] = 1] = "0x0000000000000000000000000000000000000000";
    NullAddress[NullAddress["0x0000000000000000000000000000000000000000000000000000000000000000"] = 2] = "0x0000000000000000000000000000000000000000000000000000000000000000";
})(NullAddress = exports.NullAddress || (exports.NullAddress = {}));
function isNullAddress(key) {
    if (!key)
        return true;
    return Object.values(NullAddress).includes(key);
}
exports.isNullAddress = isNullAddress;
exports.EthCoinIndex = 60;
exports.UnclaimedDomainResponse = {
    addresses: {},
    meta: {
        owner: null,
        type: '',
        ttl: 0,
    },
};
/**
 * @deprecated Use UnclaimedDomainResponse instead (deprecated since 0.3.4)
 */
exports.UNCLAIMED_DOMAIN_RESPONSE = exports.UnclaimedDomainResponse;
